common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/UsbSmartcardCertBasedAuthChallengeHandler.java

```diff
@@ -0,0 +1,149 @@
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+package com.microsoft.identity.common.internal.ui.webview.challengehandlers;
+
+import android.app.Activity;
+import android.os.Build;
+import android.webkit.ClientCertRequest;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.RequiresApi;
+
+import com.microsoft.identity.common.R;
+import com.microsoft.identity.common.java.opentelemetry.CertBasedAuthTelemetryHelper;
+import com.microsoft.identity.common.logging.Logger;
+
+/**
+ * Handles a received ClientCertRequest by prompting the user to choose from certificates
+ *  stored on a smartcard device connected via USB.
+ */
+public class UsbSmartcardCertBasedAuthChallengeHandler extends AbstractSmartcardCertBasedAuthChallengeHandler {
+
+    /**
+     * Creates new instance of UsbSmartcardCertBasedAuthChallengeHandler.
+     * A manager for smartcard CBA is retrieved, and discovery for USB devices is started.
+     * @param activity current host activity.
+     * @param usbSmartcardCertBasedAuthManager AbstractUsbSmartcardCertBasedAuthManager instance.
+     * @param dialogHolder DialogHolder instance.
+     * @param telemetryHelper CertBasedAuthTelemetryHelder instance.
+     */
+    public UsbSmartcardCertBasedAuthChallengeHandler(@NonNull final Activity activity,
+                                                     @NonNull final AbstractUsbSmartcardCertBasedAuthManager usbSmartcardCertBasedAuthManager,
+                                                     @NonNull final DialogHolder dialogHolder,
+                                                     @NonNull final CertBasedAuthTelemetryHelper telemetryHelper) {
+        super(activity, usbSmartcardCertBasedAuthManager, dialogHolder, telemetryHelper, UsbSmartcardCertBasedAuthChallengeHandler.class.getSimpleName());
+        final String methodTag = TAG + ":UsbSmartcardCertBasedAuthChallengeHandler";
+        mSmartcardCertBasedAuthManager.setConnectionCallback(new AbstractSmartcardCertBasedAuthManager.IConnectionCallback() {
+            @Override
+            public void onCreateConnection() {
+                //Reset DialogHolder to null if necessary.
+                //In this case, DialogHolder would be an ErrorDialog if not null.
+                mDialogHolder.dismissDialog();
```

Comment:
would this code ever be triggered? 

If we ever land here, I wonder what would the user see (and later proceed).

Reply:
This will be triggered when a user either cancels or fails out of CBA and sees that "authentication failed" screen with an error dialog on top. If the user plugs in again here, without closing the dialog manually, then we'll close the dialog for them. 
This code doesn't do much, but I think it improves the user exp just a little bit